compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -549,21 +557,250 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
+    ///   --> $DIR/static-impl-obligation.rs:169:20
+    ///    |
+    /// LL |     impl dyn Foo + 'static where Self: 'static {
+    ///    |                    ^^^^^^^             ^^^^^^^
+    /// LL |         fn hello(&self) where Self: 'static {}
+    ///    |                                     ^^^^^^^
+    /// ```
+    #[allow(rustc::diagnostic_outside_of_impl)]
+    #[allow(rustc::untranslatable_diagnostic)]
+    fn explain_impl_static_obligation(
+        &self,
+        diag: &mut Diag<'_>,
+        ty: Ty<'tcx>,
+        call_span: Span,
+        outlived_fr: RegionVid,
+    ) {
+        let tcx = self.infcx.tcx;
+        let ty::FnDef(def_id, args) = ty.kind() else {
+            return;
+        };
+        let Ok(Some(instance)) = ty::Instance::resolve(
+            tcx,
+            self.param_env,
+            *def_id,
+            self.infcx.resolve_vars_if_possible(args),
+        ) else {
+            return;
+        };
+        debug!(?instance);
+        let def_id = instance.def_id();
+        let mut parent = tcx.parent(def_id);
+        debug!(?def_id, ?parent);
+        let trait_preds: Vec<_> = match tcx.def_kind(parent) {
+            hir::def::DefKind::Impl { .. } => tcx
+                .trait_id_of_impl(parent)
+                .map_or(vec![], |id| elaborate_predicates_of(tcx, id).collect()),
+            hir::def::DefKind::Trait => {
+                let Some(ty) = args.get(0).and_then(|arg| arg.as_type()) else {
+                    return;
+                };
+                let mut impls = vec![];
+                tcx.for_each_relevant_impl(parent, ty, |id| {
+                    impls.push(id);
+                });
+                if let [def_id] = impls[..] {
+                    // The method we have is on the trait, but for `parent` we want to analyze the
+                    // relevant impl instead.
+                    let preds = elaborate_predicates_of(tcx, parent);
+                    parent = def_id;
+                    preds.collect()
+                } else {
+                    return;
+                }
+            }
+            _ => return,
+        };
+        debug!(?def_id, ?parent);
+        let ty = tcx.type_of(parent).instantiate_identity();
+        debug!(?ty);
+        if self.to_error_region(outlived_fr) != Some(tcx.lifetimes.re_static) {
+            return;
+        }
+        // FIXME: there's a case that's yet to be handled: `impl dyn Trait + '_ where Self: '_`
+        // causes *two* errors to be produded, one about `where Self: '_` not being allowed,
+        // and the regular error with no additional information about "lifetime may not live
+        // long enough for `'static`" without mentioning where it came from. This is because
+        // our error recovery fallback is indeed `ReStatic`. We should at some point introduce
+        // a `ReError` instead to avoid this and other similar issues.
```

Comment:
Huh, I thought we'd use [`hir::LifetimeName::Error`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.LifetimeName.html#variant.Error) when encountering `where Self: '_` and later lower it via `rbv::ResolvedArg::Error` to `RegionKind::ReError` or sth. like that, haven't checked.

Reply:
I hadn't realized (or forgot) that we had that now.

Reply:
After rebasing the issue mentioned in the FIXME no longer happens.